### PR TITLE
Added new datasource for listing instance profiles for a role

### LINKED ISSUE #23646: data source: aws_iam_instance_profiles
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I've found myself working with a module that creates an IAM role and IAM Instance Profile; however, it returns only an IAM role ARN and I need both the role and the instance profile. The AWS CLI implements a `list-instance-profiles-for-role` method which gives me the information I need, and I think it would be a great addition to Terraform.

### New or Affected Resource(s)

* data.aws_iam_instance_profiles

### Potential Terraform Configuration

```hcl
module "thing1" {
  arg1 = val1 
}

data "aws_iam_instance_profiles" "iam1" {
  role_name = module.thing1.iam_role_name
}

resource "ec2_instance" "thing2" {
  iam_instance_profile {
    name = data.aws_iam_instance_profiles.iam1.names[0]
  }
}
```

### References

- https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/iam#Client.ListInstanceProfilesForRole

### PR DESCRIPTION
**Pull request by Aegon95**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #23646

Sample Output for datasource:
image

Output from acceptance testing:

```
$ make testacc TESTS=TestAccXXX PKG=ec2

...
```

### CODE REVIEW

File: internal/service/iam/instance_profiles_data_source_test.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+resource "aws_iam_instance_profile" "test" {
+  name = %[2]q
+  role = aws_iam_role.test.name
+  path = "/testpath/"
+}
+
+data "aws_iam_instance_profiles" "test" {
+	role_name = aws_iam_role.test.name
+	depends_on = [aws_iam_role.test, aws_iam_instance_profile.test]
+}
```
Comment (kamilturek) [inline internal/service/iam/instance_profiles_data_source_test.go:56]: terrafmt fails here.
https://github.com/hashicorp/terraform-provider-aws/runs/6182456694?check_suite_focus=true

File: internal/service/iam/instance_profiles_data_source_test.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-provider-aws/internal/acctest"
+)
+
+func TestAccIAMInstanceProfilesDataSource_basic(t *testing.T) {
+	datasourceName := "data.aws_iam_instance_profiles.test"
+	resourceName := "aws_iam_instance_profile.test"
+
+	roleName := fmt.Sprintf("tf-acc-ds-instance-profile-role-%d", sdkacctest.RandInt())
+	profileName := fmt.Sprintf("tf-acc-ds-instance-profile-%d", sdkacctest.RandInt())
```
Comment (kamilturek) [inline internal/service/iam/instance_profiles_data_source_test.go:18]: We may use `sdkacctest.RandomWithPrefix` instead as it seems to be a better practice.
https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/running-and-writing-acceptance-tests.md#randomized-naming

File: internal/service/iam/instance_profiles_data_source.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		return fmt.Errorf("no IAM instance profiles found for role %s", roleName)
+	}
+
+	instanceProfiles := resp.InstanceProfiles
+
+	var arns, paths, names []string
+	for _, profile := range instanceProfiles {
+		arns = append(arns, *profile.Arn)
+		paths = append(paths, *profile.Path)
+		names = append(names, *profile.InstanceProfileName)
```
Comment (kamilturek) [inline internal/service/iam/instance_profiles_data_source.go:65]:
```suggestion
		arns = append(arns, aws.StringValue(profile.Arn))
		paths = append(paths, aws.StringValue(profile.Path))
		names = append(names, aws.StringValue(profile.InstanceProfileName))
```

File: internal/service/iam/instance_profiles_data_source.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	d.SetId(aws.StringValue(&roleName))
```
Comment (kamilturek) [inline internal/service/iam/instance_profiles_data_source.go:68]:
```suggestion
	d.SetId(roleName)
```

File: .changelog/23646.txt [hunk 5]
```diff
@@ -0,0 +1,3 @@
+```release-note:new-data-source
```
Comment (kamilturek) [inline .changelog/23646.txt:1]: The name of the changelog file should be a PR number, so `24423.txt` in this case.
https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/pullrequest-submission-and-lifecycle.md#changelog-format

File: internal/service/iam/instance_profiles_data_source.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+func DataSourceInstanceProfiles() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceInstanceProfilesRead,
+
+		Schema: map[string]*schema.Schema{
+			"role_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
```
Comment (kamilturek) [inline internal/service/iam/instance_profiles_data_source.go:21]: We might validate the max length (64) here.
https://docs.aws.amazon.com/IAM/latest/APIReference/API_ListInstanceProfilesForRole.html

**Review by kamilturek — COMMENTED** [review]
Thank you for your contribution 👍 Looks cool. I left a few ideas for improvement.

**Comment by Aegon95** [comment]
[kamilturek](https://github.com/kamilturek) I did all the changes mentioned, have a look and let me know if those are good

for role_name valdiation i have reused validation function from role resource creation (role.go), not sure if thats okay

File: internal/service/iam/instance_profiles_data_source.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	}
+}
+
+func dataSourceInstanceProfilesRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*conns.AWSClient).IAMConn
+
+	roleName := d.Get("role_name").(string)
+
+	req := &iam.ListInstanceProfilesForRoleInput{
+		RoleName: &roleName,
```
Comment (kamilturek) [inline internal/service/iam/instance_profiles_data_source.go:48]:
```suggestion
		RoleName: aws.String(roleName),
```

File: internal/service/iam/instance_profiles_data_source_test.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+				),
+			},
+		},
+	})
+}
+
+func testAccInstanceProfilesDataSourceConfig(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_iam_role" "test" {
+  name               = "%[1]s-role"
```
Comment (kamilturek) [inline internal/service/iam/instance_profiles_data_source_test.go:42]:
```suggestion
  name               = %[1]q
```

File: internal/service/iam/instance_profiles_data_source_test.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  assume_role_policy = "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Effect\":\"Allow\",\"Principal\":{\"Service\":[\"ec2.amazonaws.com\"]},\"Action\":[\"sts:AssumeRole\"]}]}"
+}
+
+resource "aws_iam_instance_profile" "test" {
+  name = "%[1]s-instance-profile"
```
Comment (kamilturek) [inline internal/service/iam/instance_profiles_data_source_test.go:47]:
```suggestion
  name = %[1]q
```

**Review by kamilturek — APPROVED** [review]
LGTM in general 👍 Thanks! 🚀

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTS=TestAccIAMInstanceProfilesDataSource_ PKG=iam
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/iam/... -v -count 1 -parallel 20 -run='TestAccIAMInstanceProfilesDataSource_'  -timeout 180m
=== RUN   TestAccIAMInstanceProfilesDataSource_basic
=== PAUSE TestAccIAMInstanceProfilesDataSource_basic
=== CONT  TestAccIAMInstanceProfilesDataSource_basic
--- PASS: TestAccIAMInstanceProfilesDataSource_basic (18.27s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/iam	26.279s
```

**Comment by ewbankkit** [comment]
@Aegon95 Thanks for the contribution :tada: :clap:.

**Pull request merged**